### DISCUSSION TITLE
How to open protocol development to a wider audience?

### DISCUSSION THREAD

**Discussion by jakmeier** [near/nearcore#8811] [General]
## Problem
It is very hard for non-Pagoda employees to work on protocol features because a great deal of expertise is required.

## Situation today
To give a bit more context, as a full-time protocol engineer employed by Pagoda, I have spent a lot of time learning about NEAR protocol and its technical details. Today, I feel I can work with most parts of the nearcore code base quite efficiently and when I look at parts of the code base that are beyond me, at least I know who to ask about it. But getting to this point required onboaring me over multiple months and a lot of help from my team colleagues.

With this background, I can now work on implementing new protocol features. Of course, I have to propose them as a NEP just like anybody else, because we want the community to be in power.

But now I want to ask: Does the community really have the tools to change the protocol? The group of people who were never part of Pagoda but are technically capable of making protocol changes seems rather small indeed.

To underline my point, let's look at approved NEPs that changed the protocol (as opposed to e.g. smart contract/wallet/other standards outside the protocol itself) and that weren't exclusively authored by Pagoda employees. I can only find [NEP-364](https://github.com/near/NEPs/blob/master/neps/nep-0364.md) and [NEP-366](https://github.com/near/NEPs/blob/master/neps/nep-0366.md) meeting that criteria. And even in these two cases, I personally played a not insignificant role to get the code to the stage where it could be accepted as reference implementation, and making sure the implementation eventually gets merged into nearcore. So even those weren't 100% independent from Pagoda.

Should we be surprised about that? I mean, how are contributors outside of Pagoda expected to onboard?
Today, the best they can do is jumping into our [Zulip instance](near.zulipchat.com/) and ask questions there. Or they can start issues and discussions on Github and we will try to help them there.

This kind of works, but it's not perfect. Contributors don't have a good way of figuring out where to ask questions. And even when they find the right place and do ask us questions, they often get only direct answers to the specific question, whereas they really need some kind of broader onboarding, or at least someone looking at their idea more broadly and guiding them in the right direction.

## Possible solution

I was thinking we could improve on this by offering an official mentorship program. Contributors who want to work on a protocol NEP could describe their idea and then ask if a mentor is available to help them get onboarded. The protocol community group, in collaboration with Pagoda, could try to find such a mentor that will help them get started.

I imagine at the start, most (if not all) mentors will be from Pagoda. But if successful, this would produce more and more contributors that are capable of mentoring another.

What do you all think? Do you agree with the problem? Do you see other solutions? And if by chance you are reading this and you would be interested in such a mentorship, please let us know!

**Comment by fadeevab** [comment]
@jakmeier It's a nice idea! What I'm thinking about: you'd need to prepare some guidelines to be ready for mentoring, and it means that you could publish those guidelines, and in turn, it means that maybe online documentation would be an easier first step. :)

**Reply by jakmeier** [reply]
Hm, we already make an effort to keep online documentation around nearcore development as up-to-date as possible. See https://near.github.io/nearcore/

I know it's far from perfect but it does contain coding style guidelines etc. 

What kind of documentation would you expect us to add to become ready for mentoring?

**Comment by as1ndu** [comment]
I have been learning Rust and  learning about the Near Protocol for sometime now.

I would really want to get involved with the project some way but sadly don't feel competent enough.

If this mentorship program kicks off. I would be honoured to be part of it so I can improve my knowledge of the Near Protocol.

**Comment by jakmeier** [comment]
Update on this topic: We had a discussion in the working group, as well as internal discussions inside Pagoda.

What we observed lately is that the interest of getting protocol features into nearcore by external contributors went down since last year when we had several ongoing efforts that required support. Hence, it doesn't look like right now is good time to introduce something like a mentorship program. We should first see natural interest before we put more effort into this.
